File: rust/bf.rs
use std::env;
use std::io;
use std::io::Read;
use std::fs::File;

struct Tape {
    pointer: usize,
    tape: [u8; 30000],
}

impl Tape {
    fn get_data(&mut self) -> u8 {
        self.tape[self.pointer]
    }

    fn increment_pointer(&mut self) {
        self.pointer += 1;
    }

    fn decrement_pointer(&mut self) {
        self.pointer -= 1;
    }

    fn increment_cell(&mut self) {
        self.tape[self.pointer] += 1;
    }

    fn decrement_cell(&mut self) {
        self.tape[self.pointer] -= 1;
    }

    fn print_cell(&mut self) {
        print!("{}", self.get_data() as u8 as char);
    }

    fn input_to_cell(&mut self) {
        self.tape[self.pointer] = io::stdin()
            .bytes()
            .next()
            .and_then(|result| result.ok())
            .map(|byte| byte as u8)
            .expect("Failed to read byte");
    }
}

#[derive(Clone, Copy, Debug)]
enum Token {
    IncrementPointer,
    DecrementPointer,
    IncrementCell,
    DecrementCell,
    PrintCell,
    Input,
    LoopStart,
    LoopEnd,
}

#[derive(Clone, Debug)]
enum Command {
    Do(Token),
    Loop(Vec<Command>),
}

fn tokenize(data: Vec<u8>) -> Vec<Token> {
    let mut tokens = Vec::new();

    for u8_char in data {
        tokens.push(match u8_char as char {
            '>' => Token::IncrementPointer,
            '<' => Token::DecrementPointer,
            '+' => Token::IncrementCell,
            '-' => Token::DecrementCell,
            '.' => Token::PrintCell,
            ',' => Token::Input,
            '[' => Token::LoopStart,
            ']' => Token::LoopEnd,
            _  => continue,
        })
    }
    tokens
}

fn parse(start: usize, tokens: &Vec<Token>) -> (Vec<Command>, usize) {
    let mut program = Vec::new();
    let mut index = start;
    while index < tokens.len() {
        match tokens[index] {
            Token::LoopStart => {
                let res = parse(index+1, &tokens);
                program.push(Command::Loop(res.0));
                index = res.1;
            }
            Token::LoopEnd => { return (program, index); }
            _ => program.push(Command::Do(tokens[index])),
        }
        index += 1;
    }
    (program, index)
}

fn run(program: &Vec<Command>, tape: &mut Tape) {
    for command in program {
        match *command {
            Command::Do(Token::IncrementPointer) => tape.increment_pointer(),
            Command::Do(Token::DecrementPointer) => tape.decrement_pointer(),
            Command::Do(Token::IncrementCell) => tape.increment_cell(),
            Command::Do(Token::DecrementCell) => tape.decrement_cell(),
            Command::Do(Token::PrintCell) => tape.print_cell(),
            Command::Do(Token::Input) => tape.input_to_cell(),
            Command::Loop(ref vec) => {
                while tape.get_data() != 0 {
                    run(vec, tape);
                }
            },
            _ => continue,
        }
    }
}

fn read_file(name: String) -> Vec<u8> {
    let mut f = File::open(name).expect("Couldn't open file");
    let mut contents = Vec::new();
    f.read_to_end(&mut contents).expect("Couldn't read file");
    contents
}

fn main() {
    let f = read_file(env::args().nth(1).expect("Need a file name"));
    let tokens = tokenize(f);
    let program = parse(0, &tokens).0;
    let mut tape = Tape { pointer: 0, tape: [0; 30000] };
    run(&program, &mut tape);
}

File: nim/bf.nim
import os, strutils

type
  Tape = ref object
    cells: array[30_000, uint8]
    head: int

  Token = enum
    IncrementHead, DecrementHead
    IncrementCell, DecrementCell
    Print, Read
    LoopStart, LoopEnd

  CommandKind = enum
    Do, Loop

  Command = object
    case kind: CommandKind
    of Do:
      tk: Token
    of Loop:
      cs: seq[Command]

proc incrementHead(t: Tape) =
  t.head = t.head + 1

proc decrementHead(t: Tape) =
  t.head = t.head - 1

proc incrementCell(t: Tape) =
  t.cells[t.head] = t.cells[t.head] + 1

proc decrementCell(t: Tape) =
  t.cells[t.head] = t.cells[t.head] - 1

proc print(t: Tape) =
  stdout.write(chr(t.cells[t.head]))

proc read(t: Tape) =
  stdout.write("Enter a char: ")
  t.cells[t.head] = uint8(parseInt(stdin.readLine()))

func getData(t: Tape): uint8 =
  return t.cells[t.head]

proc tokenize(data: string): seq[Token] =
  var tokens: seq[Token]
  for c in data:
    case c:
      of '>': tokens.add(IncrementHead)
      of '<': tokens.add(DecrementHead)
      of '+': tokens.add(IncrementCell)
      of '-': tokens.add(DecrementCell)
      of '.': tokens.add(Print)
      of ',': tokens.add(Read)
      of '[': tokens.add(LoopStart)
      of ']': tokens.add(LoopEnd)
      else:   continue
  return tokens

proc parse(tokens: seq[Token], index: int): (seq[Command], int) =
  var commands: seq[Command]
  var index = index
  while index < tokens.len:
    case tokens[index]:
      of LoopStart:
        let res = parse(tokens, index+1)
        index = res[1]
        let loop = Command(kind: Loop, cs: res[0])
        commands.add(loop)
      of LoopEnd:
        return (commands, index)
      else:
        let command = Command(kind: Do, tk: tokens[index])
        commands.add(command)
    index.inc
  return (commands, index)

proc run(program: seq[Command], tape: Tape) =
  for command in program:
    case command.kind:
      of Do:
        case command.tk:
          of IncrementHead: tape.incrementHead()
          of DecrementHead: tape.decrementHead()
          of IncrementCell: tape.incrementCell()
          of DecrementCell: tape.decrementCell()
          of Print:         tape.print()
          of Read:          tape.read()
          else:             discard
      of Loop:
        while tape.getData() != 0:
          run(command.cs, tape)

when isMainModule:
  let brainfuck = paramStr(1).readFile()
  let program = tokenize(brainfuck).parse(0)[0]
  let tape = Tape()
  run(program, tape)
